Mathlib/AlgebraicTopology/SimplicialSet/AnodyneExtensions/PairingCore.lean

```diff
@@ -0,0 +1,226 @@
+/-
+Copyright (c) 2025 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.AlgebraicTopology.SimplicialSet.AnodyneExtensions.Pairing
+
+/-!
+# Helper structure in order to construct pairings
+
+In this file, we introduce a helper structure `Subcomplex.PairingCore`
+in order to construct a pairing for a subcomplex of a simplicial set.
+The main differences with `Subcomplex.Pairing` are that we provide
+an index type `ι` in order to parametrize type (I) and type (II) simplices,
+and that the dimensions of these are definitionally `d` or `d + 1`.
+
+-/
+
+universe v u
+
+open CategoryTheory Simplicial
+
+namespace SSet.Subcomplex
+
+variable {X : SSet.{u}} (A : X.Subcomplex)
+
+/-- An helper structure in order to construct a pairing for a subcomplex of a
+simplicial set. The main differences with `Pairing` are that we provide
+an index type `ι` in order to parametrize type (I) and type (II) simplices,
+and that the dimensions of these are definitionally `d` or `d + 1`. -/
+structure PairingCore where
+  /-- the index type -/
+  ι : Type v
+  /-- the dimension of a type (II) simplex -/
+  d (s : ι) : ℕ
+  /-- a type (I) simplex -/
+  simplex (s : ι) : X _⦋d s + 1⦌
+  /-- the corresponding type (II) is the `1`-codimensional face given by this index -/
+  index (s : ι) : Fin (d s + 2)
+  nonDegenerate₁ (s : ι) : simplex s ∈ X.nonDegenerate _
+  nonDegenerate₂ (s : ι) : X.δ (index s) (simplex s) ∈ X.nonDegenerate _
+  notMem₁ (s : ι) : simplex s ∉ A.obj _
+  notMem₂ (s : ι) : X.δ (index s) (simplex s) ∉ A.obj _
+  injective_type₁' {s t : ι} (h : S.mk (simplex s) = S.mk (simplex t)) : s = t
+  injective_type₂' {s t : ι}
+    (h : S.mk (X.δ (index s) (simplex s)) = S.mk (X.δ (index t) (simplex t))) : s = t
+  type₁_ne_type₂' (s t : ι) : S.mk (simplex s) ≠ S.mk (X.δ (index t) (simplex t))
+  surjective' (x : A.N) :
+    ∃ (s : ι), x.toS = S.mk (simplex s) ∨ x.toS = S.mk (X.δ (index s) (simplex s))
+
+variable {A}
+
+/-- The `PairingCore` structure induced by a pairing. The opposite construction
+is `PairingCore.pairing`. -/
+noncomputable def Pairing.pairingCore (P : A.Pairing) [P.IsProper] :
+    A.PairingCore where
+  ι := P.II
+  d s := s.1.dim
+  simplex s := ((P.p s).1.cast (P.isUniquelyCodimOneFace s).dim_eq).simplex
+  index s := (P.isUniquelyCodimOneFace s).index rfl
+  nonDegenerate₁ s := ((P.p s).1.cast (P.isUniquelyCodimOneFace s).dim_eq).nonDegenerate
+  nonDegenerate₂ s := by
+    rw [(P.isUniquelyCodimOneFace s).δ_index rfl]
+    exact s.1.nonDegenerate
+  notMem₁ s := ((P.p s).1.cast (P.isUniquelyCodimOneFace s).dim_eq).notMem
+  notMem₂ s := by
+    rw [(P.isUniquelyCodimOneFace s).δ_index rfl]
+    exact s.1.notMem
```

Comment:
Side note: `SSet.Subcomplex.N.notMem` should be renamed to `SSet.Subcomplex.N.not_mem`